association.go

```diff
@@ -393,18 +401,40 @@ func (a *Association) sendCookieEcho() error {
 	return nil
 }
 
-// Close ends the SCTP Association and cleans up any state
-func (a *Association) Close() error {
+// Shutdown initiates the shutdown sequence. The returned channel will be
+// closed once the connection is closed. When an error is returned, the
+// channel will be nil.
+func (a *Association) Shutdown() (done <-chan struct{}, err error) {
 	a.log.Debugf("[%s] closing association..", a.name)
 
-	a.setState(closed)
+	state := a.getState()
 
-	err := a.netConn.Close()
+	if state != established {
+		return nil, fmt.Errorf("%w: shutdown %s", errShutdownNonEstablished, a.name)
+	}
 
-	a.closeAllTimers()
+	// Attempt a graceful shutdown.
+	a.setState(shutdownPending)
 
-	// awake writeLoop to exit
-	a.closeWriteLoopOnce.Do(func() { close(a.closeWriteLoopCh) })
+	a.lock.Lock()
+
+	if a.inflightQueue.size() == 0 {
+		// No more outstanding, send shutdown.
+		a.willSendShutdown = true
+		a.awakeWriteLoop()
+		a.setState(shutdownSent)
+	}
+
+	a.lock.Unlock()
+
+	return a.closeWriteLoopCh, nil
```

Comment:
I am a bit nervous about returning the channel.  May I hear your thoughts on this?

Reply:
I think I just wanted a way for the caller to know when the shutdown process has completed since it is does not happen in an instant like `Close` does.

Reply:
Shutdown sequence (esp. graceful one)  is always a headache... 
I personally feel that the Shutdown() should block on the channel by itself.  If necessary, caller can always choose to call Shutdown() from a goroutine for async operation. Hmm, let me ask Sean.

@Sean-Der What is thought on this? The Shutdown() method triggers the shutdown sequence. Should it block until the sequence is complete, or return a channel so that the caller could choose to wait on the channel...?
Pion mixes Javascript (non-blocking) and Go (blocking) styles which confuses myself sometimes...

Reply:
I think blocking is the way to go. You can optionally pass in a context to be notified. I am not an expert Go developer either though :)

* [crypto/tls](https://golang.org/pkg/crypto/tls/#Conn.Close) is a blocking Close
* HTTP provides two APIs [Context](https://golang.org/pkg/net/http/#Server.Shutdown) and [Blocking](https://golang.org/pkg/net/http/#Server.Close)
* [net.Conn](https://golang.org/pkg/net/#Conn) is blocking as well.

---

So excited to have you both back I missed working together! I learn so much from your PRs and so fun to see Pion progress so quickly without me :)

Reply:
Thanks, I like the `context.Context` way. I just pushed an update that modifies the `sctp.Association.Shutdown`'s signature to:

```go
// Shutdown initiates the shutdown sequence. The method blocks until the
// shutdown sequence is completed and the connection is closed, or until the
// passed context is done, in which case the context's error is returned.
func (a *Association) Shutdown(ctx context.Context) error {
```

Does that work?

Reply:
> So excited to have you both back I missed working together! I learn so much from your PRs and so fun to see Pion progress so quickly without me :)

Aww thanks :heart: I miss Pion, I learned so much from the whole community and I don't think I'd have the clients I currently have if it weren't for you guys!